File: local_bot/spotify_client.py
"""Client Spotify : OAuth, recherche, dédoublonnage et ajout playlist."""

from __future__ import annotations

import json
import logging
import re
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

import spotipy
from spotipy.oauth2 import SpotifyOAuth

logger = logging.getLogger(__name__)

SCOPES = (
    "playlist-modify-public "
    "playlist-modify-private "
    "playlist-read-private"
)

FEAT_PATTERN = re.compile(
    r"\s*[\(\[]?\s*(feat\.?|ft\.?|featuring|avec|&)\s+.+?[\)\]]?\s*$",
    re.IGNORECASE,
)
PUNCT_PATTERN = re.compile(r"[^\w\s]", re.UNICODE)


class SpotifyClientError(Exception):
    """Erreur liée à l'API Spotify."""


def normalize_query_part(value: str) -> str:
    cleaned = FEAT_PATTERN.sub("", value)
    cleaned = PUNCT_PATTERN.sub(" ", cleaned)
    cleaned = re.sub(r"\s+", " ", cleaned).strip()
    return cleaned


def _fingerprint(artist: str, title: str) -> str:
    return f"{artist.casefold().strip()}|{title.casefold().strip()}"


class SpotifyPlaylistSync:
    def __init__(
        self,
        *,
        client_id: str,
        client_secret: str,
        redirect_uri: str,
        playlist_id: str,
        cache_path: Path,
        token_cache_path: Path | None = None,
    ) -> None:
        self.playlist_id = playlist_id
        self.cache_path = cache_path
        self._playlist_uris: set[str] = set()
        self._status_by_nrj_id: dict[str, str] = {}
        self._status_by_fp: dict[str, str] = {}
        self._cache_entries: list[dict[str, Any]] = []

        auth_manager = SpotifyOAuth(
            client_id=client_id,
            client_secret=client_secret,
            redirect_uri=redirect_uri,
            scope=SCOPES,
            cache_path=str(token_cache_path or Path(".cache")),
            open_browser=True,
        )
        self.sp = spotipy.Spotify(auth_manager=auth_manager)

        self._load_local_cache()
        self._refresh_playlist_uris()

    def _load_local_cache(self) -> None:
        if not self.cache_path.exists():
            return
        try:
            raw = json.loads(self.cache_path.read_text(encoding="utf-8"))
        except (OSError, json.JSONDecodeError) as exc:
            logger.warning("Cache local illisible (%s), réinitialisation.", exc)
            return

        entries = raw if isinstance(raw, list) else raw.get("tracks", [])
        if not isinstance(entries, list):
            return

        self._cache_entries = [e for e in entries if isinstance(e, dict)]
        for entry in self._cache_entries:
            nrj_id = entry.get("nrj_song_id")
            uri = entry.get("uri")
            status = str(entry.get("status") or "already")
            artist = str(entry.get("artist") or "")
            title = str(entry.get("title") or "")
            if nrj_id:
                self._status_by_nrj_id[str(nrj_id)] = status
            if uri:
                self._playlist_uris.add(str(uri))
            if artist and title:
                self._status_by_fp[_fingerprint(artist, title)] = status

    def _save_local_cache(self) -> None:
        self.cache_path.parent.mkdir(parents=True, exist_ok=True)
        payload = {
            "updated_at": datetime.now(timezone.utc).isoformat(),
            "tracks": self._cache_entries,
        }
        self.cache_path.write_text(
            json.dumps(payload, ensure_ascii=False, indent=2),
            encoding="utf-8",
        )

    def _refresh_playlist_uris(self) -> None:
        logger.info("Chargement des titres existants de la playlist…")
        offset = 0
        limit = 100
        while True:
            try:
                page = self.sp.playlist_items(
                    self.playlist_id,
                    fields="items.track.uri,next",
                    additional_types=["track"],
                    limit=limit,
                    offset=offset,
                )
            except spotipy.SpotifyException as exc:
                raise SpotifyClientError(
                    f"Impossible de lire la playlist : {exc}"
                ) from exc

            for item in page.get("items") or []:
                track = item.get("track") or {}
                uri = track.get("uri")
                if uri:
                    self._playlist_uris.add(uri)

            if not page.get("next"):
                break
            offset += limit

        logger.info("%d URI(s) déjà dans la playlist.", len(self._playlist_uris))

    def has_seen_nrj_song(self, nrj_song_id: str) -> bool:
        return str(nrj_song_id) in self._status_by_nrj_id

    def is_handled_successfully(
        self,
        nrj_song_id: str,
        artist: str = "",
        title: str = "",
    ) -> bool:
        """True si déjà ajouté / confirmé en playlist (pas un vieux not_found)."""
        status = self._status_by_nrj_id.get(str(nrj_song_id))
        if status in ("added", "already"):
            return True
        if artist and title:
            fp_status = self._status_by_fp.get(_fingerprint(artist, title))
            if fp_status in ("added", "already"):
                return True
        return False

    def search_track(self, artist: str, title: str) -> dict[str, Any] | None:
        artist_q = normalize_query_part(artist)
        title_q = normalize_query_part(title)
        queries = [
            f'track:"{title_q}" artist:"{artist_q}"',
            f"{title_q} {artist_q}",
            f"track:{title_q}",
        ]

        for query in queries:
            try:
                results = self.sp.search(q=query, type="track", limit=5)
            except spotipy.SpotifyException as exc:
                raise SpotifyClientError(f"Recherche Spotify échouée : {exc}") from exc

            items = (results.get("tracks") or {}).get("items") or []
            if not items:
                continue

            artist_lower = artist_q.lower()
            for item in items:
                names = " ".join(
                    a.get("name", "") for a in (item.get("artists") or [])
                ).lower()
                if artist_lower and artist_lower.split()[0] in names:
                    return item
            return items[0]

        return None

    def add_track(
        self,
        *,
        nrj_song_id: str,
        artist: str,
        title: str,
    ) -> str:
        """
        Returns:
            "added" | "already" | "not_found"
        """
        nrj_song_id = str(nrj_song_id)

        if self.is_handled_successfully(nrj_song_id, artist, title):
            logger.info("Déjà traité (cache) : %s - %s", artist, title)
            return "already"

        track = self.search_track(artist, title)
        if not track:
            logger.warning("Introuvable sur Spotify : %s - %s", artist, title)
            # not_found reste réessayable (pas « handled successfully »)
            self._remember(nrj_song_id, None, artist, title, status="not_found")
            return "not_found"

        uri = track["uri"]
        track_name = track.get("name", title)
        track_artists = ", ".join(
            a.get("name", "") for a in (track.get("artists") or [])
        )

        if uri in self._playlist_uris:
            logger.info(
                "Déjà dans la playlist : %s - %s", track_artists, track_name
            )
            self._remember(nrj_song_id, uri, artist, title, status="already")
            return "already"

        try:
            # spotipy récent utilise POST /playlists/{id}/items
            self.sp.playlist_add_items(self.playlist_id, [uri])
        except spotipy.SpotifyException as exc:
            raise SpotifyClientError(
                f"Échec de l'ajout à la playlist : {exc}"
            ) from exc

        self._playlist_uris.add(uri)
        self._remember(nrj_song_id, uri, artist, title, status="added")
        logger.info(
            "Ajouté avec succès : %s - %s (%s)", track_artists, track_name, uri
        )
        return "added"

    def _remember(
        self,
        nrj_song_id: str,
        uri: str | None,
        artist: str,
        title: str,
        *,
        status: str,
    ) -> None:
        self._status_by_nrj_id[nrj_song_id] = status
        if artist and title:
            self._status_by_fp[_fingerprint(artist, title)] = status
        if uri:
            self._playlist_uris.add(uri)
        entry = {
            "nrj_song_id": nrj_song_id,
            "uri": uri,
            "artist": artist,
            "title": title,
            "status": status,
            "added_at": datetime.now(timezone.utc).isoformat(),
        }
        self._cache_entries.append(entry)
        self._save_local_cache()

File: src/SpotifyClient.php
<?php

declare(strict_types=1);

final class SpotifyClientError extends RuntimeException
{
}

/**
 * Client Spotify : refresh token, recherche, dédoublonnage, ajout playlist.
 */
final class SpotifyClient
{
    private const SCOPES = 'playlist-modify-public playlist-modify-private playlist-read-private';
    private const TOKEN_URL = 'https://accounts.spotify.com/api/token';
    private const API_BASE = 'https://api.spotify.com/v1';
    private const TIMEOUT = 20;

    private string $accessToken = '';
    private int $expiresAt = 0;

    public function __construct(
        private readonly string $clientId,
        private readonly string $clientSecret,
        private readonly string $redirectUri,
        private readonly string $playlistId,
        private readonly string $tokenPath,
        private readonly Cache $cache,
        private readonly ?Logger $logger = null,
    ) {
        $this->ensureAccessToken();
        $this->refreshPlaylistUris();
    }

    public static function scopes(): string
    {
        return self::SCOPES;
    }

    public static function buildAuthorizeUrl(
        string $clientId,
        string $redirectUri,
        string $state
    ): string {
        return 'https://accounts.spotify.com/authorize?' . http_build_query([
            'client_id'     => $clientId,
            'response_type' => 'code',
            'redirect_uri'  => $redirectUri,
            'scope'         => self::SCOPES,
            'state'         => $state,
            'show_dialog'   => 'true',
        ]);
    }

    /**
     * Échange le code OAuth contre des tokens et les enregistre.
     * Spotify omet parfois refresh_token en ré-auth : on conserve l’ancien s’il existe.
     *
     * @return array<string, mixed>
     */
    public static function exchangeCode(
        string $clientId,
        string $clientSecret,
        string $redirectUri,
        string $code,
        string $tokenPath
    ): array {
        $payload = http_build_query([
            'grant_type'   => 'authorization_code',
            'code'         => $code,
            'redirect_uri' => $redirectUri,
        ]);

        $tokens = self::tokenRequest($clientId, $clientSecret, $payload);

        $newRefresh = $tokens['refresh_token'] ?? null;
        if (!is_string($newRefresh) || $newRefresh === '') {
            $existing = self::readTokenFile($tokenPath);
            $oldRefresh = $existing['refresh_token'] ?? null;
            if (is_string($oldRefresh) && $oldRefresh !== '') {
                $tokens['refresh_token'] = $oldRefresh;
            }
        }

        self::saveTokens($tokenPath, $tokens);

        $savedRefresh = $tokens['refresh_token'] ?? null;
        if (!is_string($savedRefresh) || $savedRefresh === '') {
            throw new SpotifyClientError(
                'Spotify n’a pas renvoyé de refresh_token. Révoquez l’accès de l’app sur '
                . 'https://www.spotify.com/account/apps/ puis rouvrez auth.php '
                . '(show_dialog force le consentement).'
            );
        }

        return $tokens;
    }

    /**
     * Vérifie que le dossier data/ est créable/inscriptible.
     */
    public static function assertTokenPathWritable(string $tokenPath): void
    {
        $dir = dirname($tokenPath);
        if (!is_dir($dir)) {
            if (!@mkdir($dir, 0755, true) && !is_dir($dir)) {
                throw new SpotifyClientError(
                    'Impossible de créer le dossier data/ (' . $dir . '). '
                    . 'Vérifiez les permissions du serveur (utilisateur PHP / www).'
                );
            }
        }
        if (!is_writable($dir)) {
            throw new SpotifyClientError(
                'Le dossier data/ n’est pas inscriptible (' . $dir . '). '
                . 'Corrigez les permissions (chmod/chown) pour l’utilisateur PHP.'
            );
        }
        if (is_file($tokenPath) && !is_writable($tokenPath)) {
            throw new SpotifyClientError(
                'Le fichier token.json existe mais n’est pas inscriptible (' . $tokenPath . ').'
            );
        }
    }

    /**
     * Indique si token.json contient un refresh_token non vide.
     */
    public static function hasRefreshToken(string $tokenPath): bool
    {
        $tokens = self::readTokenFile($tokenPath);
        $refresh = $tokens['refresh_token'] ?? null;
        return is_string($refresh) && $refresh !== '';
    }

    public static function normalizeQueryPart(string $value): string
    {
        $cleaned = preg_replace(
            '/\s*[\\(\\[]?\\s*(feat\\.?|ft\\.?|featuring|avec|&)\\s+.+?[\\)\\]]?\\s*$/iu',
            '',
            $value
        ) ?? $value;
        $cleaned = preg_replace('/[^\w\s]/u', ' ', $cleaned) ?? $cleaned;
        $cleaned = preg_replace('/\s+/u', ' ', $cleaned) ?? $cleaned;
        return trim($cleaned);
    }

    public function hasSeenNrjSong(string $nrjSongId): bool
    {
        return $this->cache->hasSeenNrjSong($nrjSongId);
    }

    /**
     * @return 'added'|'already'|'not_found'
     */
    public function addTrack(string $nrjSongId, string $artist, string $title): string
    {
        if ($this->cache->hasSeenNrjSong($nrjSongId)) {
            $this->logger?->info('Déjà traité (cache NRJ) : ' . $artist . ' - ' . $title);
            return 'already';
        }

        $track = $this->searchTrack($artist, $title);
        if ($track === null) {
            $this->logger?->warning('Introuvable sur Spotify : ' . $artist . ' - ' . $title);
            $this->cache->remember($nrjSongId, null, $artist, $title, 'not_found');
            return 'not_found';
        }

        $uri = (string) ($track['uri'] ?? '');
        $trackName = (string) ($track['name'] ?? $title);
        $trackArtists = implode(', ', array_map(
            static fn(array $a): string => (string) ($a['name'] ?? ''),
            is_array($track['artists'] ?? null) ? $track['artists'] : []
        ));

        if ($uri === '' || $this->cache->hasUri($uri)) {
            $this->logger?->info('Déjà dans la playlist : ' . $trackArtists . ' - ' . $trackName);
            $this->cache->remember($nrjSongId, $uri !== '' ? $uri : null, $artist, $title, 'already');
            return 'already';
        }

        $this->apiRequest('POST', '/playlists/' . rawurlencode($this->playlistId) . '/tracks', [
            'uris' => [$uri],
        ]);

        $this->cache->addUri($uri);
        $this->cache->remember($nrjSongId, $uri, $artist, $title, 'added');
        $this->logger?->info(
            'Ajouté avec succès : ' . $trackArtists . ' - ' . $trackName . ' (' . $uri . ')'
        );
        return 'added';
    }

    /**
     * @return array<string, mixed>|null
     */
    public function searchTrack(string $artist, string $title): ?array
    {
        $artistQ = self::normalizeQueryPart($artist);
        $titleQ = self::normalizeQueryPart($title);

        $queries = [
            'track:"' . $titleQ . '" artist:"' . $artistQ . '"',
            $titleQ . ' ' . $artistQ,
            'track:' . $titleQ,
        ];

        foreach ($queries as $query) {
            $results = $this->apiRequest('GET', '/search?' . http_build_query([
                'q'     => $query,
                'type'  => 'track',
                'limit' => 5,
            ]));

            $items = $results['tracks']['items'] ?? [];
            if (!is_array($items) || $items === []) {
                continue;
            }

            $artistLower = str_lower($artistQ);
            $firstWord = $artistLower !== '' ? explode(' ', $artistLower)[0] : '';

            foreach ($items as $item) {
                if (!is_array($item)) {
                    continue;
                }
                $names = '';
                foreach ($item['artists'] ?? [] as $a) {
                    if (is_array($a)) {
                        $names .= ' ' . str_lower((string) ($a['name'] ?? ''));
                    }
                }
                if ($firstWord !== '' && str_contains($names, $firstWord)) {
                    return $item;
                }
            }

            $first = $items[0];
            return is_array($first) ? $first : null;
        }

        return null;
    }

    private function refreshPlaylistUris(): void
    {
        $this->logger?->info('Chargement des titres existants de la playlist…');
        $offset = 0;
        $limit = 100;
        $count = 0;

        while (true) {
            $page = $this->apiRequest(
                'GET',
                '/playlists/' . rawurlencode($this->playlistId) . '/items?' . http_build_query([
                    'fields'            => 'items.track.uri,next',
                    'additional_types'  => 'track',
                    'limit'             => $limit,
                    'offset'            => $offset,
                ])
            );

            $items = $page['items'] ?? [];
            if (is_array($items)) {
                $uris = [];
                foreach ($items as $item) {
                    if (!is_array($item)) {
                        continue;
                    }
                    $uri = $item['track']['uri'] ?? null;
                    if (is_string($uri) && $uri !== '') {
                        $uris[] = $uri;
                        $count++;
                    }
                }
                $this->cache->mergeUris($uris);
            }

            if (empty($page['next'])) {
                break;
            }
            $offset += $limit;
        }

        $this->logger?->info($count . ' URI(s) déjà dans la playlist.');
    }

    private function ensureAccessToken(): void
    {
        $tokens = $this->loadTokens();
        $now = time();

        if (
            isset($tokens['access_token'], $tokens['expires_at'])
            && is_string($tokens['access_token'])
            && (int) $tokens['expires_at'] > $now + 60
        ) {
            $this->accessToken = $tokens['access_token'];
            $this->expiresAt = (int) $tokens['expires_at'];
            return;
        }

        $refresh = $tokens['refresh_token'] ?? null;
        if (!is_string($refresh) || $refresh === '') {
            throw new SpotifyClientError(
                'Aucun refresh_token. Ouvrez auth.php dans le navigateur pour autoriser l’app.'
            );
        }

        $payload = http_build_query([
            'grant_type'    => 'refresh_token',
            'refresh_token' => $refresh,
        ]);

        $fresh = self::tokenRequest($this->clientId, $this->clientSecret, $payload);
        if (!isset($fresh['refresh_token'])) {
            $fresh['refresh_token'] = $refresh;
        }
        self::saveTokens($this->tokenPath, $fresh);

        $this->accessToken = (string) $fresh['access_token'];
        $this->expiresAt = time() + (int) ($fresh['expires_in'] ?? 3600);
    }

    /**
     * @return array<string, mixed>
     */
    private function loadTokens(): array
    {
        return self::readTokenFile($this->tokenPath);
    }

    /**
     * @return array<string, mixed>
     */
    public static function readTokenFile(string $tokenPath): array
    {
        if (!is_file($tokenPath)) {
            return [];
        }
        $raw = file_get_contents($tokenPath);
        if ($raw === false || $raw === '') {
            return [];
        }
        $decoded = json_decode($raw, true);
        return is_array($decoded) ? $decoded : [];
    }

    /**
     * @param array<string, mixed> $tokens
     */
    public static function saveTokens(string $tokenPath, array $tokens): void
    {
        self::assertTokenPathWritable($tokenPath);

        $refresh = $tokens['refresh_token'] ?? null;
        if (!is_string($refresh) || $refresh === '') {
            $existing = self::readTokenFile($tokenPath);
            $old = $existing['refresh_token'] ?? null;
            if (is_string($old) && $old !== '') {
                $tokens['refresh_token'] = $old;
            }
        }

        $expiresIn = (int) ($tokens['expires_in'] ?? 3600);
        $payload = [
            'access_token'  => (string) ($tokens['access_token'] ?? ''),
            'refresh_token' => (string) ($tokens['refresh_token'] ?? ''),
            'token_type'    => (string) ($tokens['token_type'] ?? 'Bearer'),
            'scope'         => (string) ($tokens['scope'] ?? self::SCOPES),
            'expires_in'    => $expiresIn,
            'expires_at'    => time() + $expiresIn,
            'obtained_at'   => gmdate('c'),
        ];

        $json = json_encode($payload, JSON_UNESCAPED_UNICODE | JSON_PRETTY_PRINT);
        if ($json === false) {
            throw new SpotifyClientError('Impossible d’encoder token.json');
        }
        if (@file_put_contents($tokenPath, $json . "\n", LOCK_EX) === false) {
            throw new SpotifyClientError(
                'Impossible d’écrire token.json (' . $tokenPath . '). '
                . 'Vérifiez les permissions du dossier data/.'
            );
        }
    }

    /**
     * @return array<string, mixed>
     */
    private static function tokenRequest(
        string $clientId,
        string $clientSecret,
        string $payload
    ): array {
        try {
            $res = Http::request('POST', self::TOKEN_URL, $payload, [
                'Content-Type'  => 'application/x-www-form-urlencoded',
                'Authorization' => 'Basic ' . base64_encode($clientId . ':' . $clientSecret),
            ], self::TIMEOUT);
        } catch (Throwable $e) {
            throw new SpotifyClientError('Échec token Spotify : ' . $e->getMessage(), 0, $e);
        }

        $data = json_decode($res['body'], true);
        if (!is_array($data) || $res['status'] >= 400 || empty($data['access_token'])) {
            $msg = is_array($data)
                ? (string) ($data['error_description'] ?? $data['error'] ?? $res['body'])
                : $res['body'];
            throw new SpotifyClientError('Token Spotify HTTP ' . $res['status'] . ' : ' . $msg);
        }

        return $data;
    }

    /**
     * @param array<string, mixed>|null $jsonBody
     * @return array<string, mixed>
     */
    private function apiRequest(string $method, string $path, ?array $jsonBody = null): array
    {
        if ($this->accessToken === '' || time() >= $this->expiresAt - 30) {
            $this->ensureAccessToken();
        }

        $url = str_starts_with($path, 'http') ? $path : self::API_BASE . $path;
        $headers = [
            'Authorization' => 'Bearer ' . $this->accessToken,
            'Accept'        => 'application/json',
        ];
        $body = null;
        if ($jsonBody !== null) {
            $encoded = json_encode($jsonBody);
            if ($encoded === false) {
                throw new SpotifyClientError('JSON invalide pour requête Spotify');
            }
            $headers['Content-Type'] = 'application/json';
            $body = $encoded;
        }

        try {
            $res = Http::request($method, $url, $body, $headers, self::TIMEOUT);
        } catch (Throwable $e) {
            throw new SpotifyClientError('Échec API Spotify : ' . $e->getMessage(), 0, $e);
        }

        if ($res['status'] === 204 || $res['body'] === '') {
            return [];
        }

        $data = json_decode($res['body'], true);
        if ($res['status'] >= 400) {
            $msg = is_array($data)
                ? json_encode($data['error'] ?? $data, JSON_UNESCAPED_UNICODE)
                : $res['body'];
            throw new SpotifyClientError('API Spotify HTTP ' . $res['status'] . ' : ' . $msg);
        }

        return is_array($data) ? $data : [];
    }
}
